scripts/saml-configuration

```diff
@@ -0,0 +1,154 @@
+    if 'metadata_url' not in idps[idp] or 'entity_id' not in idps[idp]:
+        logging.error('No \'metadata_url\' and/or \'entity_id\' field for IdP '
+                      '\'%s\' in the SAML configuration', idp)
+        continue
+
+    # Get the data from the metadata url of the IdP.
+    data: Dict[str, Any] = Parser.parse_remote(
+        idps[idp]['metadata_url'],
+        validate_cert = True,
+        entity_id = idps[idp]['entity_id']
+    )
+
+    if 'idp' not in data:
+        logging.error('Cannot receive metadata for IdP %s', idp)
+        logging.debug(data)
+        continue
+
+    # Some documentation:
+    # - https://github.com/onelogin/python3-saml
+    # - https://github.com/onelogin/python3-saml/blob/master/src\
+    #     /onelogin/saml2/idp_metadata_parser.py
+
+    if 'x509cert' in data['idp']:
+        # A single certificate, used for both signing and encryption.
+        certs_list.append(data['idp']['x509cert'])
+    elif ('x509certMulti' in data['idp']
+          and 'signing' in data['idp']['x509certMulti']):
+        # There are multiple certificates, maybe different ones for
+        # signing and encryption.
+        certs_list.extend(data['idp']['x509certMulti']['signing'])
+
+    if not certs_list:
+        logging.error('Cannot find any certificates for IdP %s', idp)
+        continue
+
+    # Format the certificates properly.
+    # See https://github.com/onelogin/python3-saml/blob/master/src\
+    #   /onelogin/saml2/utils.py
+    certs: Iterator[str] = map(Utils.format_cert, certs_list)
+
+    # Map the certificates to their x509 objects in order to be able to
+    # determine the validity of the certificate.
+    # See https://cryptography.io/en/latest/x509/reference.html\
+    #   #cryptography.x509.Certificate
+    certs_objects: Iterator[Tuple[str, x509.Certificate]] = (
+        (cert, x509.load_pem_x509_certificate(cert.encode(), backend=default_backend()))
+        for cert in certs
+    )
+
+    # Check for the best certificate available:
+    #  - It is currently valid.
+    #  - It is valid for as long as possible.
+    try:
+        new_cert, new_cert_object = max(
+            (
+                (cert, cert_object)
+                for cert, cert_object in certs_objects
+                if is_valid(cert_object)
+            ),
+            key=lambda t: t[1].not_valid_after,
+        )
+    except ValueError:
+        logging.error('Cannot find any valid certificate for IdP %s', idp)
+        logging.debug('', exc_info=True)
+        continue
+
+    cert_path: str = idp_cert_path_template.format(idp)
+
+    # Write the certificate to the appropriate location.
+    try:
+        with open(cert_path, mode = 'w') as f:
+            f.write(new_cert)
+    except OSError:
+        logging.error('Cannot write certificate for IdP %s to %s', idp, cert_path)
+        continue
+
+    # Set the appropriate permissions for the certificate file.
+    os.chmod(cert_path, stat.S_IRUSR | stat.S_IWUSR)
```

Comment:
Certificates are public, right? But if it is important for these to be secret, the permissions need to be applied at creation with `os.umask()` or `os.open(mode=…)`, not later.

Reply:
Hm, I consulted the [Zulip SAML docs](https://zulip.readthedocs.io/en/stable/production/authentication-methods.html#saml).
They assign 644 to the public certificates: `find /etc/zulip/saml/ -type f -exec chmod 644 -- {} +`. So I think we should do that here, too. I'll adapt the mode.
But of course the mode is to be applied in a proper manner. Thank you for the hint, I'm sorry that I didn't consider this. I'll fix this.

Reply:
Done. Do you think that we should change the permissions if the file(s) already exist(s)? Or should we rely on the one who created the file that they knew what they do?

Reply:
Perhaps we should do a prompt if a file with different permissions is detected? "There is already a certificate file under <path> with non-standard permissions XXX. Do you want to change the permissions to the default 644? [Y/n]"

Reply:
I don't know... I think the script will often run non-interactively (i.e. during installation, as cron-job, etc.). Maybe we could do some logging instead (on `INFO` level)?